### ISSUE TITLE
will there ever be a way to use an actual Google account

### ISSUE THREAD

**Issue by iiimuuij** [iBicha/playlet#283]
I know the app is about privacy but I really want to have all of my recommendations

**Comment by iBicha** [comment]
Thanks for creating this issue. I know this is partially about recommendations, but I will put some points about bringing your Youtube profile into Playlet in general, in case anyone else is looking at this info.

### Using Google account directly
This something that many people ask about, but might not be as simple.

Connecting to a Google account and reading subscriptions, recommendations, etc would require Playlet to have its own "extractor". Extractors are what reads data from YouTube through various methods, and are not easy to maintain. Usually there is continuous breakage to fix, and also changes happen in an A/B tested manner, where half the people can't reproduce the issues. This makes it hard to maintain. Playlet being a newer app, with currently just myself as the developer, it is hard for me to commit to such big project. Invidious not only provides a layer of privacy, but it is also a layer of stability shielding Playlet for YouTube continuously changing data format and breaking things.

The other thing I need to study before I even consider going this route is how Roku would react, and how it would see this from a ToS point of view. Roku is a company that favors big corporations that brings in big revenue. If YouTube were to complain about Playlet connecting to Google accounts, Roku might side with Google, and I wouldn't be surprised to try justifying banning Playlet from the store. At that point, side-loading the app is remains the only option. And even then, Roku can also ban developer mode on certain devices if they detect you're trying to install an app they don't approve of. It's a very strict platform, so I need to understand the implications of this before doing anything.

Overall, is it possible? Yes. Do I think we will have this in Playlet anytime soon (e.g. in the next few months)? I don't think so.

### Importing from YouTube to Invidious
The next best thing is importing data from YouTube profile to Invidious. While Invidious has a guide on how to [import your subscriptions](https://docs.invidious.io/export-youtube-subscriptions/), for some people this might not be enough.

This is why I started working on [yt2alt](https://github.com/iBicha/yt2alt), which makes it easy to import profile, including recommendations.
Right now, the tool is interactive, and exports recommended videos as a Playlist. I plan to add a [non-interactive mode](https://github.com/iBicha/yt2alt/issues/1) which can make bringing data from YouTube to Invidious automated. Once that's done, then we can schedule a sync between YouTube and Invidious (say every few hours) and then your recommendations would follow you to Playlet.

Adding a small detail here, before creating [yt2alt](https://github.com/iBicha/yt2alt), I used a [script](https://github.com/iBicha/playlet/blob/b076517c1b1bb86f369deef2a6db0fc67950afd0/tools/profile-sync.js) (which uses [yt-dlp](https://github.com/yt-dlp/yt-dlp)) that I run every 6 hours to import my profile into Invidious. Unfortunately this thing is currently broken, that's why I'm trying to bring this functionality to `yt2alt` as an alternative that anyone can use.

### Casting from YouTube mobile app
The other thing I'm working on to reduce the friction is using YouTube mobile app to cast directly to Playlet. This is an experiment https://github.com/iBicha/playlet/pull/276 that I think will become a feature of Playlet, and it should make it nicer to find content you relate to. I've done enough prototyping to be convinced this is something that should be part of the app, and I'm working to finish it.

### Making a built-in importer in Playlet
Another possible thing to do is to make a tool (similar to yt2alt) but built into Playlet. Since we're leveraging [Youtube.js](https://github.com/LuanRT/YouTube.js) as an extractor, this importer can be added to the Playlet web app. This feature might look like this:
- You use the web app to help you login to Google account
- We read your profile data
- And take that data and send it to your Invidious profile
This feature has a couple of downsides
- It would increase the app size significantly, and might lead to slower loading times
- I need to also consider the implications of adding "Login to YouTube" to Playlet.
This feature, because of these two limitations, might need to be implemented as a separate plugin, as opposed to a built-in functionality.


I hope this clarifies things for anyone wondering about using their YouTube profile. I'm committed to trying to find ways that make using your existing profile less tedious.

**Comment by Sebastin3244** [comment]
How do I create an invidious account?

**Comment by iBicha** [comment]
> How do I create an invidious account?

You can create an Invidious account on your own instance (some people call it an instance, others call it a server) if you have one, or use one of the [public instances](https://api.invidious.io/). A profile is instance specific, and does not sync automatically between instances.

From Playlet, you can go to the profile page, and choose login:
![profiles](https://raw.githubusercontent.com/iBicha/playlet/83d0b707b5c96f901acf018a260d42c925e5a6d7/docs/screenshots/profiles.jpg)

And follow the instructions. You will be able to create an Invidious profile on your currently selected Invidious instance and link it to Playlet.

**Comment by iBicha** [comment]
tl;dr
- This feature is possible to implement
- I've made some progress, but it's not finished yet. I don't have a release date.
- I've included developer instructions if you want to check out what was done so far.

---------

I want to provide a little update, since it's been over a year since this issue was created.
This is by far the most requested feature, and I have not abandoned this yet.

It's actually possible to login using YouTube account, and I have tested this and was able to login, show subscriptions, playlists, and watch history straight from YouTube.
While the proof of concept confirms it's possible using OAuth, there are many things to be addressed, such as (but not limited):
- Management of life cycle of authorization (access token, refresh token, etc)
- Marking videos as watched, so they can appear in the watch history
- Architectural issues introduced by this feature, which led the app to dead-lock if the web app is used.
- Managing more abstraction differences: Playlet was designed with Invidious in mind after all. Some things need to be redone.
- Concerns over securely storying YouTube's access token, so malicious users can't access your YouTube account. Some mitigations include [encryption](https://github.com/iBicha/playlet/pull/539), and explaining the risks when using Playlet in dev mode. 

All of these things need to be addressed, and a lot of thorough testing need to be done to avoid issues.
It's currently hard for me to find the time to finish all of this, so I do not have an ETA, but I know I will get to it at some point.

For those curious:
If you setup Playlet for development using the [documentation](https://github.com/iBicha/playlet/blob/main/docs/README.md), then you test this proof of concept yourself.

I put this change behind a "feature flag" here:
https://github.com/iBicha/playlet/blob/4a8b12294d66d0436471b2709a75f5eda027939e/playlet-lib/src/manifest#L16

By switching `YT_OAUTH_SUPPORT` to `true`, the login button prompts a dialog to choose which service to login with (YouTube or Invidious), and just follow the instructions.
The issues/limitations I mentioned earlier still apply.

**Comment by iBicha** [comment]
YouTube account support now available in Canary release
